File: templates/typescript/interfaces/IAgent.ts
import { AgentRequest }   from './IAgentRequest';
import { AgentResponse }  from './IAgentResponse';
import { DiscoveryEntry } from './IAgentDiscovery';

/**
 * Sentrix agent interface.
 * Every Sentrix agent must implement this contract.
 *
 * Identity note
 * ─────────────
 * `agentId` is required. `owner` is optional — a local secp256k1 key is
 * sufficient for signed ANR records and P2P discovery without an on-chain
 * wallet. ERC-8004 on-chain registration remains available as an opt-in.
 * See identity/IdentityProvider for the LocalKeystoreIdentity default.
 */
export interface IAgent {
  // ─── Identity ─────────────────────────────────────────────────────────────
  /** Sentrix agent URI, e.g. "sentrix://agent/0xABC..." */
  readonly agentId: string;
  /**
   * Owner identifier — Ethereum address when using ERC-8004 or key-derived
   * identity; any unique string otherwise. Defaults to "anonymous".
   */
  readonly owner?: string;
  /** Optional IPFS / on-chain metadata URI */
  readonly metadataUri?: string;
  /** Human-readable metadata bag */
  readonly metadata?: AgentMetadata;

  // ─── Capabilities ─────────────────────────────────────────────────────────
  /** Return the list of capability names this agent exposes */
  getCapabilities(): string[];

  // ─── Request handling ─────────────────────────────────────────────────────
  /** Primary dispatch method — all inbound calls arrive here */
  handleRequest(request: AgentRequest): Promise<AgentResponse>;
  /** Optional pre-processing hook (auth, rate-limit, logging…) */
  preProcess?(request: AgentRequest): Promise<void>;
  /** Optional post-processing hook (audit log, billing…) */
  postProcess?(response: AgentResponse): Promise<void>;

  // ─── Discovery (optional) ─────────────────────────────────────────────────
  /** Announce this agent to the discovery layer */
  registerDiscovery?(): Promise<void>;
  /** Gracefully withdraw from the discovery layer */
  unregisterDiscovery?(): Promise<void>;

  // ─── Delegation / permissions (optional) ─────────────────────────────────
  /** Return true if `caller` is permitted to invoke `capability` */
  checkPermission?(caller: string, capability: string): Promise<boolean>;

  // ─── ANR / Identity exposure ──────────────────────────────────────────────
  /**
   * Return the full ANR (Agent Network Record) for this agent.
   *
   * The ANR is the authoritative self-description of the agent on the mesh:
   * its identity, capabilities, network endpoint, and health status.
   * Callers can use this to inspect a live agent without querying the
   * discovery layer.
   */
  getAnr(): DiscoveryEntry;

  /**
   * Return the libp2p PeerId derived from this agent's secp256k1 ANR key.
   *
   * The PeerId is derived from the same key used to sign ANR records —
   * one keypair, one identity across both the ANR layer and the P2P transport.
   *
   * Returns null for anonymous agents (no signing key configured).
   */
  getPeerId(): Promise<string | null>;

  // ─── Signing (optional) ───────────────────────────────────────────────────
  /** EIP-712 compatible message signing */
  signMessage?(message: string): Promise<string>;
}

export interface AgentMetadata {
  name: string;
  version: string;
  description?: string;
  author?: string;
  license?: string;
  repository?: string;
  tags?: string[];
  resourceRequirements?: {
    minMemoryMb?: number;
    minCpuCores?: number;
    storageGb?: number;
  };
}

File: templates/python/interfaces/iagent.py
"""
Sentrix agent interface.
Every Sentrix agent must implement this abstract base class.

Identity note
─────────────
``agent_id`` is required; ``owner`` is optional.
ERC-8004 on-chain registration is not required — a local secp256k1 key is
sufficient for signing ANR records and P2P discovery.
See identity.provider.LocalKeystoreIdentity for the default no-wallet option.
"""

from abc import ABC, abstractmethod
from typing import List, Optional
from .agent_request import AgentRequest
from .agent_response import AgentResponse


class IAgent(ABC):
    # ── Identity ───────────────────────────────────────────────────────────
    agent_id: str                    # e.g. "sentrix://agent/0xABC..."
    owner: str     = "anonymous"     # Wallet address or arbitrary identifier.
                                     # Required for ERC-8004; optional otherwise.
    metadata_uri: Optional[str] = None
    metadata: Optional[dict]    = None

    # ── Capabilities ───────────────────────────────────────────────────────
    @abstractmethod
    def get_capabilities(self) -> List[str]:
        """Return list of capability names this agent exposes."""
        ...

    # ── Request handling ───────────────────────────────────────────────────
    @abstractmethod
    async def handle_request(self, request: AgentRequest) -> AgentResponse:
        """Primary dispatch — all inbound calls arrive here."""
        ...

    async def pre_process(self, request: AgentRequest) -> None:
        """Optional hook: auth, rate-limit, logging. Override as needed."""
        pass

    async def post_process(self, response: AgentResponse) -> None:
        """Optional hook: audit log, billing. Override as needed."""
        pass

    # ── Discovery (optional) ───────────────────────────────────────────────
    async def register_discovery(self) -> None:
        """Announce this agent to the discovery layer."""
        pass

    async def unregister_discovery(self) -> None:
        """Gracefully withdraw from the discovery layer."""
        pass

    # ── Delegation / permissions (optional) ────────────────────────────────
    async def check_permission(self, caller: str, capability: str) -> bool:
        """Return True if `caller` is permitted to invoke `capability`."""
        return True  # open by default; override for production

    # ── Signing (optional) ─────────────────────────────────────────────────
    async def sign_message(self, message: str) -> str:
        """EIP-712 compatible message signing. Override with your key."""
        raise NotImplementedError("sign_message not implemented")
